SO-3948: refactor/redesign RF2 import/export endpoints

cmark opened the pull request:
This PR improves the existing RF2 import/export endpoints in the following way:
* Keeps the old APIs in deprecated form in backward compatibility (few properties needed to be removed, but the way they behave is the same, still using in-memory objects to identify import/export tasks)
* The new API does not use any in-memory objects and can be safely used in scenarios when there are multiple Snow Owl instances working together on the same dataset
* New RF2 Import endpoints:
  * `POST /:path/import?createVersions=true&type=full` 
    * Initiates an import with the specified config values and a `multipart/form-data` file
    * Returns the Location header of the import task with the import ID
  * `GET /:path/import/:id` - Retrieve the status of the import task using its ID
  * `DELETE /:path/import/:id` - Delete an import task using its ID (executing on a running import will mark the task as CANCEL_REQUESTED, but it won't stop the import from finishing because the current importer implementation does not support cancellation)
* New RF2 Export endpoint:
  * `GET /:path/export?type=full`
  

Additional improvements present in this PR:
* New exception handlers added to `ControllerExceptionMapper`
* New `runAsJob` helper added to `AsyncRequest` to ease the execution of requests as a job
* Removed FeatureToggle system (superseded by jobs and locks)
* The following bundles have been merged into the snowowl.core bundle to form a true snowowl-core module:
  * snowowl.datastore
  * snowowl.identity
  * snowowl.identity.file (as default included identity provider)
  * snowowl.scripting.api
  * snowowl.scripting.groovy (as default included scripting language)
* The `snowowl.datastore.server.tests` project has been merged into the `snowowl.core.tests` project.

cmark committed:
[core] move c.b.s.core.attachments package to snowowl.core

cmark committed:
[core] merge identity module into snowowl.core

cmark committed:
[core] migrate references to identity classes

cmark committed:
[core] merge identity.file back to snowowl.core

NOOP and FILE authentication will be the default available identity
providers supported by the core plug-in.

cmark committed:
[core] migrate identity.ldap plugin

Convert LDAP fragment to plugin and use plug-in infrastructure to
register the plugin to the system.

cmark committed:
[build] remove identity and identity.file from build

cmark committed:
[build] fix reference to identity and identity.file modules

cmark committed:
SO-3948: add new RF2 export rest API

Also, remove codeSystemShortName from RF2 export Java API and infer the
relative codesystem from the current reference branch.
Support locales String in RF2 export Java API.

cmark committed:
SO-3948: remove produces part from export endpoint

Since the export endpoint produces two different media types
(octet_stream when the request succeeds and the file download starts and
application/json if there was a failure during the request execution) it
is necessary to remove the produces values otherwise Spring will not be
able to detect the error handler when an exception occurs.
Also, remove the Resource type from the ResponseEntity will remove the
JSON response example from the Swagger UI.

cmark committed:
SO-3948: add more exception handlers to ControllerExceptionMapper

cmark committed:
SO-3948: add getEffectiveTime(String, String) method to EffectiveTimes

cmark committed:
SO-3948: use String for RF2 export start/end effectiveTimes configs

cmark committed:
SO-3948: use new `/:path/export` endpoint in existing RF2 export tests

cmark committed:
SO-3948: use BranchContext instead of the `referenceBranch` property...

...and the RevisionIndexRequestBuilder to support new simplified export
REST API path syntax: `/snomed-ct/v3/SNOMEDCT-UK-CL/export`.

cmark committed:
SO-3948: update deprecated javadoc message in SnomedExportRestService

cmark committed:
SO-3948: update SnomedRf2ImportRequest and Builder

Remove unnecessary codeSystemShortName (will be computed from the
current branch context).
Support snapshot behaviour selection in RevisionIndexRequestBuilder.
Add BranchContext alternative to Features.getImportFeatureToggle().
Add new commit() method alternatives to TransactionContext.

cmark committed:
SO-3948: deprecate SnomedImportRestService and all `/imports` endpoints

cmark committed:
Merge branch 'feature/single-core-module' into...

...feature/SO-3948-remove-import-export-config-objects

Conflicts:
	core/com.b2international.snowowl.core/pom.xml
	core/com.b2international.snowowl.identity.file/META-INF/MANIFEST.MF
	core/com.b2international.snowowl.identity.file/pom.xml
	core/com.b2international.snowowl.identity.ldap/META-INF/MANIFEST.MF
	core/com.b2international.snowowl.identity/META-INF/MANIFEST.MF
	core/com.b2international.snowowl.identity/pom.xml
	core/com.b2international.snowowl.identity/src/com/b2international/snowowl/identity/domain/Permission.java
	core/com.b2international.snowowl.identity/src/com/b2international/snowowl/identity/domain/Role.java
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/EvaluateQueryRefSetMemberRequest.java
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/EvaluateQueryRefSetRequest.java
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java
	snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ImportRequest.java

cmark committed:
[core] move Concept and Concepts generic model to snowowl.core

cmark committed:
[core] move core.authorization package from datastore to core

cmark committed:
[core] move core.client package from datastore to core

cmark committed:
[core] move core.commit package from datastore to core

cmark committed:
[core] move core.repository package from datastore to core

cmark committed:
[core] move core.request package from datastore to core

cmark committed:
[core] move core.rate package from datastore to core

cmark committed:
[core] move core.validation packages from datastore to core

cmark committed:
[core] move core.merge package from datastore to core

cmark committed:
[core] move core.compare package from datastore to core

cmark committed:
[core] move core.messaging package from datastore to core

cmark committed:
[core] move core.locks and oplock packages from datastore to core

cmark committed:
[core] move datastore.config to core.config

cmark committed:
[core] move CodeSystem classes to core.codesystem

cmark committed:
[core] move datastore.converter package to core.request

cmark committed:
[core] move datastore.events package content to core.branch...

...and core.repository

cmark committed:
[core] move datastore.request.compare package to core.compare

cmark committed:
[core] renaming classes in core.branch package

cmark committed:
[core] moving review related classes to core.review package

cmark committed:
[core] move compare related classes to core.compare package

cmark committed:
[core] move branch request classes to core.branch package...

...and review request classes to core.review package

cmark committed:
[core] move datastore.review classes to core.review package

cmark committed:
[core] move datastore.request classes to core.request package

cmark committed:
[core] move datastore.repository and datastore.system to...

...core.repository and core.system

cmark committed:
[core] move datastore.file to core.attachments

Also, move datastore.index to core.repository

cmark committed:
[core] remove unused classes

cmark committed:
[core] move codesystem and codesystem version related classes to...

core.codesystem and core.codesystem.version

cmark committed:
[core] move job related classes to core.jobs

cmark committed:
[core] move OsPropertyTester to snowowl.core

cmark committed:
[core] organize imports, fix moved class references, export packages

cmark committed:
[core] move content availability info provider to core.repository

cmark committed:
[core] remove snowowl.datastore plugin

cmark committed:
[core] use FunctionalInterface for ContentAvailabilityInfoProvider

cmark committed:
[core] depend on snowowl.core instead of datastore

cmark committed:
[core] merge datastore.server.tests into core.tests

cmark committed:
[core] merge scripting.api and scripting.groovy into snowowl.core

cmark committed:
[core] remove leftover identity project files

cmark committed:
[core] remove bundles from server.core.feature

cmark committed:
SO-3948: move import config validation into SnomedRf2ImportRequest

cmark committed:
SO-3948: remove datastore from core aggregator pom.xml

cmark committed:
SO-3948: fix snomed.core.feature dependencies

Remove scripting.groovy reference from validation.snomed module

cmark committed:
SO-3948: support CodeSystem lookup from a reference branch

Useful in import/export APIs where you need to create versions during
the import for the closest CodeSystem compared to the target branch.

cmark committed:
SO-3948: remove unused XML dependencies from core.rest module

cmark committed:
SO-3948: update launch configurations

cmark committed:
SO-3948: use class path scanning for ScriptEngine registry init

cmark committed:
SO-3948: use class path scanning for validation issue detail extension..

...registry init

cmark committed:
SO-3948: remove invalid module references from dist project

cmark committed:
SO-3948: support runaAsJob in AsyncRequest

This allows API clients to run any request as a job easily without much
boilerplate code.
Example that imports RF2 as a job using the RF2 Import Request API.
```
  SnomedRequests.rf2().prepareImport()
		.setRf2ArchiveId(rf2ArchiveId)
		.setReleaseType(Rf2ReleaseType.FULL)
		.setCreateVersions(true)
		.build("SNOMEDCT/HEAD")
		.runAsJob(importJobId, "Importing SNOMED CT RF2 file")
		.execute(bus)
		.getSync(1, TimeUnit.MINUTES);
```

cmark committed:
SO-3948: extract SHA-1 ID generation from a unique source value to IDs

cmark committed:
SO-3948: add CodeSystemURI.head() helper method

cmark committed:
SO-3948: add deprecation notice to old RF2 import configuration object

cmark committed:
SO-3948: add rf2ArchiveId, type and createVersions to JSON rep...

...of SnomedRf2ImportRequest

cmark committed:
SO-3948: add new /:path/import endpoints

Use new endpoint in all current RF2 import test cases.

cmark committed:
SO-3948: add new exception handlers to ControllerExceptionMapper

Use RestControllerAdvice to remove ResponseBody annotation from return
types.
Add MultipartException handler.
Add HttpRequestMethodNotSupportedException handler.

cmark committed:
SO-3948: fix failed RF2 import API test cases

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/branch/review/ConceptChangesGetRequestBuilder.java:
```diff
@@ -34,7 +33,7 @@
 	@Override
 	protected Request<RepositoryContext, ConceptChanges> doBuild() {
-		return new ReadConceptChangesRequest(reviewId);	
+		return new ReviewGetConceptChangesRequest(reviewId);	
```
This class (`ConceptChangesGetRequestBuilder`) is the only one in its package that hasn't been renamed to include a `Review` prefix and match the request class name. 

Please rename it to `ReviewGetConceptChangesRequestBuilder` for consistency! 🙂

cmark replied:
It was on my TODO list, but forgot to do it, thanks for reminding me. 😅

apeteri commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/id/IDs.java:
```diff
@@ -17,20 +17,35 @@
+	 * @param value
+	 * @return
+	 */
+	public static String sha1(String value) {
```
Please indicate in the method name that it is going to be shorter than a regular SHA-1 hash.

cmark committed:
SO-3948: export core.scripts package

cmark committed:
SO-3948: remove unused includePreReleaseContent from exporter fragments

apeteri commented on snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/SnomedRf2ExportRestService.java:
```diff
@@ -0,0 +1,111 @@
+			.setRefSetExportLayout(params.getRefSetLayout() == null ? globalExportLayout : params.getRefSetLayout())
+			.build(this.repositoryId, branch)
+			.execute(getBus())
+			.getSync();
```
Can we wait for the results here? What is the currently configured timeout value?

cmark replied:
No timeout value at the moment, we wait indefinitely.
We can best guess a proper timeout value based on the selected export configuration, but not sure if that's a good idea. Not sure if we can convert this to an async method either, my guess is that Spring needs the HTTP thread to finish the file download work.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java:
```diff
@@ -286,8 +262,8 @@ public ExportResult execute(final RepositoryContext context) {
 			
-			final long effectiveTimeStart = startEffectiveTime != null ? startEffectiveTime.getTime() : 0;
-			final long effectiveTimeEnd =  endEffectiveTime != null ? endEffectiveTime.getTime() : Long.MAX_VALUE;
+			final long effectiveTimeStart = startEffectiveTime != null ? EffectiveTimes.getEffectiveTime(startEffectiveTime, DateFormats.SHORT) : 0;
```
Please move parsing to `SnomedRf2ExportRequestBuilder#doBuild`, if possible, so that we know that all parameters are valid by the time the request starts.

cmark replied:
Good idea! 👍

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/importer/Rf2TransactionContext.java:
```diff
@@ -111,25 +111,32 @@ public void clearContents() {
+		// after successful commit register all commited IDs to CIS
+		final ISnomedIdentifierService cis = service(ISnomedIdentifierService.class);
+		if (cis.importSupported()) {
+			cis.register(idsToRegister);
```
Consider wrapping this part in a `try-catch` block (although not sure if we want to move forward with an import that has desynced from the CIS).

cmark replied:
Well, yeah, that's something that should not happen, otherwise duplicate content will start appearing and that's always a bad sign. The CIS registration logic has been simplified greatly to allow as much ID as possible in, so I think we are safe for now. 😄

cmark committed:
SO-3948: rename ConceptChangesGetRequestBuilder to...

...ReviewGetConceptChangesRequestBuilder

cmark committed:
SO-3948: convert start/end effective times during request build time

cmark committed:
SO-3948: use dedicated method for short SHA-1 generation

Also add a full SHA-1 hash method to IDs.

cmark committed:
SO-3948: convert start/end effective times during request build time...

...(missing bits)

cmark committed:
SO-3948: convert jobId to jobKey

Supplied id is no longer acts as a unique key, but as a basis to
generate the unique SHA-1 key of a remote job.
This will ensure that the unique key of the job is URL safe and can be
returned safely from REST API endpoints.
Add filterByKey/filterByKeys method to SearchJobRequestBuilder.
Support 'restart' option in ScheduleJobRequests. With this option the
caller can restart (replace in fact) an already
completed/failed/canceled job with the same key easily (will eliminate a
lot of boilerplate code around clients).

cmark committed:
SO-3948: migrate classification services to new job keys

cmark committed:
SO-3948: migrate SNOMED CT APIs to new job keys

cmark committed:
SO-3948: migrate validation APIs to new job keys

Also, move validation endpoints to `/snomed-ct/v3` where they actually
belong.

cmark committed:
SO-3948: fix incorrect job ID used in progress monitor init

cmark committed:
SO-3948: remove feature toggle support

Superseded by unique job keys and lock objects.
Add `optionalService(Class)` method to ServiceProvider.
Add `isJobRunning(String)` method to ServiceProvider.

cmark committed:
SO-3948: support optionalService method in DelegatingContext...

...and not in all current ServiceProvider implementations

cmark committed:
SO-3948: use Strings instead of enums for release type and refset layout
